workshop1/slides/src/slides.md

```diff
@@ -0,0 +1,664 @@
+  , averageAge   :: Double
+  } deriving Show
+```
+
+* Add a notion of failure
+
+```haskell
+type DataModel = Maybe Candidate
+```
+
+* Create a list of candidates
+
+```haskell
+type DataModel = [Maybe Candidate]
+```
+
+
+
+
+# Data Transformation
+
+## Goal
+
+* Write some functions to:
+  * Convert CSV into data model
+  * Convert data model into JSON
+* Demonstrate power of `Generic`
+* Introduce `map`
+
+### Approach
+
+* Write a function `parseLine :: CSVLine -> Maybe Candidate`
+* Write a function `parseFile :: CSV -> DataModel`
+* Generate a function `encode :: DataModel -> JSON`
+* Apply these to every row
+
+![](images/data-transformation.pdf)
+
+
+## Writing `parseLine :: CSVLine -> Maybe Candidate`
+
+```haskell
+type CSVLine = String -- type alias
+
+parseLine :: CSVLine -> Maybe Candidate
+parseLine line = makeCandidate (undefined)
+
+makeCandidate :: [String] -> Maybe Candidate
+makeCandidate fields = undefined
+
+fromMaybeAge
+  :: (Double -> Candidate)
+  -> Maybe Double
+  -> Maybe Candidate
+fromMaybeAge partialCandidate maybeAge = undefined
+```
+
+* `_` means "for every other value"
+* `splitOn` splits a string into a list of strings on a character
+* `readMaybe` decodes a double from a string (may fail)
+
+
+## Writing `parseLine :: CSVLine -> Maybe Candidate`
+
+Our solution:
+
+```haskell
+parseLine :: CSVLine -> Maybe Candidate
+parseLine line = makeCandidate (splitOn ',' line)
+
+makeCandidate :: [String] -> Maybe Candidate
+makeCandidate [c, p, s, a] =
+  fromMaybeAge (Candidate c p s) (readMaybe a)
+makeCandidate _ = Nothing
+
+fromMaybeAge
+  :: (Double -> Candidate)
+  -> Maybe Double
+  -> Maybe Candidate
+fromMaybeAge partialCandidate Nothing  = Nothing
+fromMaybeAge partialCandidate (Just a) = Just (partialCandidate a)
+```
+
+
+## Stripping Quotes
+
+```haskell
+stripQ :: String -> String
+stripQ = undefined
+  where
+    leftStrip :: String -> String
+    leftStrip ('"' : xs) = xs
+    leftStrip xs         = xs
+
+    rightStrip :: String -> String
+    rightStrip = undefined
+```
+
+* Use `leftStrip` and `reverse` to implement `rightStrip`
+* Use `leftStrip` and `rightStrip` to implement `stripQ`
+* `reverse` reverses items in a list
+* `where` used to define local functions
+
+## Stripping Quotes
+
+Our solution:
+```haskell
+stripQ :: String -> String
+stripQ = leftStrip . rightStrip
+  where
+    leftStrip :: String -> String
+    leftStrip ('"' : xs) = xs
+    leftStrip xs         = xs
+
+    rightStrip :: String -> String
+    rightStrip = reverse . leftStrip . reverse
+```
+
+* Point-free style: composition FTW!
+
+## Improving `parseLine`
+
+* Using `stripQ` to improve `makeCandidate`
+
+```haskell
+...
+
+makeCandidate :: [String] -> Maybe Candidate
+makeCandidate [c, p, s, a] =
+  let c' = stripQ c
+      p' = stripQ p
+      s' = stripQ s
+  in fromMaybeAge (Candidate c' p' s') (readMaybe a)
+makeCandidate _ = Nothing
+
+...
+```
+
+* `let ... in ...` binding useful to split up functions
```

Comment:
Maybe keep on using the `where` syntax instead?

Reply:
:+1: